src/crypto_helper/algorithm.rs

```diff
@@ -361,6 +363,143 @@ pub struct ZlibInput {
+    Argon2id,
+}
+#[derive(Eq, Clone, Copy, PartialEq, Debug, Serialize, Deserialize, Default)]
+pub enum Argon2Version {
+    Version10,
+    #[default]
+    Version13,
+}
+#[derive(Eq, Clone, PartialEq, Debug, Serialize, Deserialize)]
+pub struct Argon2HashAction {
+    pub memory: u32,
+    pub iters: u32,
+    pub paralelism: u32,
+    pub output_len: usize,
+    #[serde(serialize_with = "serialize_bytes", deserialize_with = "deserialize_bytes")]
+    pub salt: Vec<u8>,
+    #[serde(serialize_with = "serialize_bytes", deserialize_with = "deserialize_bytes")]
+    pub data: Vec<u8>,
+    #[serde(serialize_with = "serialize_bytes", deserialize_with = "deserialize_bytes")]
+    pub secret: Vec<u8>,
+    pub variant: Argon2Variant,
+    pub version: Argon2Version,
+}
+
+#[derive(Eq, Clone, PartialEq, Debug, Serialize, Deserialize)]
+pub enum Argon2Action {
+    Hash(Argon2HashAction),
+    Verify(#[serde(serialize_with = "serialize_bytes", deserialize_with = "deserialize_bytes")] Vec<u8>),
+}
+#[derive(Eq, Clone, PartialEq, Debug, Serialize, Deserialize, Default)]
+pub struct Argon2Input {
+    pub action: Argon2Action,
+    #[serde(serialize_with = "serialize_bytes", deserialize_with = "deserialize_bytes")]
+    pub data: Vec<u8>,
+}
+
+impl Argon2Input {
+    pub fn set_variant(&self, variant: Argon2Variant) -> Self {
+        let mut input = self.clone();
+        if let Argon2Action::Hash(ref mut action) = input.action {
+            action.variant = variant;
+        }
+        input
+    }
+    pub fn set_version(&self, version: Argon2Version) -> Self {
+        let mut input = self.clone();
+        if let Argon2Action::Hash(ref mut action) = input.action {
+            action.version = version;
+        }
+        input
+    }
+}
+impl<'a> TryFrom<&'a str> for Argon2Version {
+    type Error = (); // todo: proper error type
+
+    fn try_from(value: &'a str) -> Result<Self, Self::Error> {
+        match value {
+            "Argon13" => Ok(Self::Version13),
+            "Argon10" => Ok(Self::Version10),
+            _ => Err(()),
+        }
+    }
+}
+impl From<Argon2Variant> for argon2::Algorithm {
+    fn from(value: Argon2Variant) -> Self {
+        match value {
+            Argon2Variant::Argon2d => argon2::Algorithm::Argon2d,
+            Argon2Variant::Argon2i => argon2::Algorithm::Argon2i,
+            Argon2Variant::Argon2id => argon2::Algorithm::Argon2id,
+        }
+    }
+}
+
+impl From<Argon2Version> for argon2::Version {
+    fn from(value: Argon2Version) -> Self {
+        match value {
+            Argon2Version::Version10 => argon2::Version::V0x10,
+            Argon2Version::Version13 => argon2::Version::V0x13,
+        }
+    }
+}
+
+impl From<&Argon2HashAction> for argon2::Params {
+    fn from(value: &Argon2HashAction) -> Self {
+        argon2::ParamsBuilder::new()
+            .m_cost(value.memory)
+            .p_cost(value.paralelism)
+            .t_cost(value.iters)
+            .build()
+            .unwrap()
```

Comment:
I know that it'll never panic, but add the message anyway:

```suggestion
            .expect("argon2::ParamsBuilder should never fail")
```

Reply:
https://github.com/TheBestTvarynka/crypto-helper/pull/47/files#r1483254713 this one is not fixed